Subject: [PATCH] Read name from '/proc/<pid>/comm' preferably
Maintainer: Miroslav Koškár <http://mkoskar.com/>
Version: v2

When running some interpreted application e.g., ranger, tmux using
'automatic-rename' will name window as 'python'. This is because
/proc/<pid>/cmdline will look like /usr/bin/python... On the other hand
/proc/<pid>/comm returns 'ranger' which corresponds to 'ps -p "$pid" -o comm='
and is what user really wants to see.

There is a limitation though, such that /proc/<pid>/comm content is truncated
to 16 chars (15 + \0). This patch takes content out of /proc/<pid>/comm always
if it exists with fallback to /proc/<pid>/cmdline with max of 32 chars.

One could devise logic, to take data from both sources and if 'comm' is
a prefix of 'cmdline' (after extracting just last bit of path) then take
'cmdline' one over 'comm'.

---
 osdep-linux.c | 24 +++++++++++-------------
 1 file changed, 11 insertions(+), 13 deletions(-)

diff --git a/osdep-linux.c b/osdep-linux.c
index ad4c11c..ad5c82d 100644
--- a/osdep-linux.c
+++ b/osdep-linux.c
@@ -27,35 +27,33 @@
 
 #include "tmux.h"
 
+#define MAXNAMELEN 32
+
 char *
 osdep_get_name(int fd, unused char *tty)
 {
 	FILE	*f;
 	char	*path, *buf;
 	size_t	 len;
-	int	 ch;
 	pid_t	 pgrp;
 
 	if ((pgrp = tcgetpgrp(fd)) == -1)
 		return (NULL);
 
-	xasprintf(&path, "/proc/%lld/cmdline", (long long) pgrp);
+	xasprintf(&path, "/proc/%lld/comm", (long long) pgrp);
 	if ((f = fopen(path, "r")) == NULL) {
 		free(path);
-		return (NULL);
+		xasprintf(&path, "/proc/%lld/cmdline", (long long) pgrp);
+		if ((f = fopen(path, "r")) == NULL) {
+			free(path);
+			return (NULL);
+		}
 	}
 	free(path);
 
-	len = 0;
-	buf = NULL;
-	while ((ch = fgetc(f)) != EOF) {
-		if (ch == '\0')
-			break;
-		buf = xrealloc(buf, len + 2);
-		buf[len++] = ch;
-	}
-	if (buf != NULL)
-		buf[len] = '\0';
+	buf = xmalloc(sizeof(char) * (MAXNAMELEN + 1));
+	len = fread(buf, sizeof(char), MAXNAMELEN, f);
+	buf[len] = '\0';
 
 	fclose(f);
 	return (buf);
